cmd/api/src/api/v2/saved_queries.go

```diff
@@ -242,3 +244,52 @@ func (s Resources) DeleteSavedQuery(response http.ResponseWriter, request *http.
 		response.WriteHeader(http.StatusNoContent)
 	}
 }
+
+// UnshareSavedQueryRequest represents the payload sent to the unshare endpoint
+type UnshareSavedQueryRequest struct {
+	UserIds []uuid.UUID `json:"user_ids"`
+	Self    bool        `json:"self"`
+}
+
+// UnshareSavedQuery allows an owner of a shared query, a user that has a saved query shared to them, or an admin, to remove sharing privileges.
+// A user who owns a query may unshare a query from anyone they have shared to
+// A user who had a query shared to them may unshare that query from themselves
+// And admins may unshare queries that have been shared to other users
+func (s Resources) UnshareSavedQuery(response http.ResponseWriter, request *http.Request) {
+	var (
+		rawSavedQueryID = mux.Vars(request)[api.URIPathVariableSavedQueryID]
+		unshareRequest  UnshareSavedQueryRequest
+	)
+
+	if user, isUser := auth.GetUserFromAuthCtx(ctx2.FromRequest(request).AuthCtx); !isUser {
+		api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, "No associated user found", request), response)
+	} else if savedQueryID, err := strconv.ParseInt(rawSavedQueryID, 10, 64); err != nil {
+		api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, api.ErrorResponseDetailsFromMalformed, request), response)
+	} else if err := json.NewDecoder(request.Body).Decode(&unshareRequest); err != nil {
+		api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, err.Error(), request), response)
+	} else {
+		if unshareRequest.Self {
+			// User is trying to unshare from themselves
+			if err := s.DB.DeleteSavedQueryPermissionsForUser(request.Context(), savedQueryID, user.ID); err != nil {
+				api.HandleDatabaseError(request, response, err)
+			}
```

Comment:
We need to verify that they aren't the query owner here correct? Otherwise that's a no-op. 

I can see why a user might want to unshare someone else sharing a query to them but you can't share your own query, it's filtered out in the `/share` handler?

Reply:
This would no-op in that situation. I can easily add a check to avoid the db call

Reply:
Added a check for if a saved query is shared to the user when attempting to unshare from themselves.
When a user creates a saved query, do not have an associated saved_query_permission for their user id and the newly created saved query id.
We also prevent a user from sharing to themselves.

This means that a user will never have a sqp for a query they created. Simply checking if a user has a sqp to that saved query allows us to check for both scenarios:

1. Ensure a user does not own the query
2. Ensure a user has the query shared to them